feat: pull components cmd

alvarosabu opened the pull request:
This PR introduces the new implementation of the `pull-components` command

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Other (please describe):

## How to test this PR

1. Clone the repo and pull the branch
2. Add several components to a Storyblok Space or use the default test one `295017` or `295018`
3. Run `pnpm install`
4. Run `pnpm run dev components pull --space <SPACE_ID>`

### Testing checklist

- [x] pull-components from a space
   - [x] should create a `components.json` file inside of `.storyblok/components/295017` directory containing an array with all the components available
   - [x] it should prompt the user with a successful message `✔ Components downloaded successfully in ./storyblok/components/295017/components.json` 
- [x] should prompt an error if flag space is missing `Please provide the space as argument --space YOUR_SPACE_ID.`
- [ ] should prompt an API error if the space ID is incorrect
- [x] should prompt an unauthorized error if the user is not logged in before operating this command
- [x] components pull in `--separate-files` 
  - [x] should create a file per component. Example:  component named `Space A Component Inside Folder 1` should generate `space-a-component-inside-folder-1.json` 
  - [x] Should prompt a warning `The --filename option is ignored when using --separate-files` if `--filename` flag is used along `--sf` 
- [x] components pull with custom `--path` 
   - [x] should generate a `components.295017.json` file inside of the path provided.
   - [x] should generate individual files inside the correspondent path if used together with `--separate-files`
- [x]  components pull with custom `--filename`
   - [x] should generate a `custom-filename.json` inside of the `.storyblok/components/295017` directory
   - [ ] should generate a `custom-filename.json` inside of the provided path if `--path` is provided
- [x] components pull with custom `--suffix`
   - [x] should generate a `components.suffix.json` inside of the `.storyblok/components/295017` directory
   - [x] should generate a `custom-filename.suffix.json` inside of the provided path if `--path` is provided
   - [x] should create a file per component with the same custom suffix if `--separate-files` is used. Example:  component named `Space A Component Inside Folder 1` should generate `space-a-component-inside-folder-1.suffix.json`
- [x]  components pull component-name
   - [x] should pull a single component if the component name is passed `storyblok components pull component-name` 
   - [x] should generate an individual `component-name.json` file inside `.storyblok/components/295017`
   - [x] should throw an error if no component-name is passed.

alvarosabu committed:
feat: basic pull-component and save to json functionality

alvarosabu committed:
feat: basic pull components actions tests

alvarosabu committed:
feat: basic tests for pull components command

alvarosabu committed:
chore: fix lint

alvarosabu committed:
chore: fix lint

alvarosabu committed:
feat: separate-files option

alvarosabu committed:
Merge branch 'next' into feature/pull-components-cmd

alvarosabu committed:
feat: suffix command and new file organization for pull-components

alvarosabu committed:
feat: add pull as as subcommand

alvarosabu committed:
feat: components presets

alvarosabu committed:
feat: pass space data as an object to the save components to files action

alvarosabu committed:
feat: simplify file saving logic by removing nested structure and making suffix optional

alvarosabu committed:
feat: update resolved path for saving components to include space directory inside

alvarosabu committed:
Merge branch 'feature/refactor-language-command-updated-dx' into feature/pull-components-cmd

alvarosabu committed:
feat: update saveComponentsToFiles function and related tests

- Updated actions to use customFetch instead of ofetch
- Modified the `saveComponentsToFiles` function to accept a structured `spaceData` object containing components, groups, and presets, instead of separate parameters.
- Updated the default filename behavior to save as `components.json` instead of including the space ID in the filename.
- Enhanced test cases to reflect changes in the function signature and filename generation logic.
- Removed unused imports and interfaces to streamline the codebase.

alvarosabu committed:
chore: remove block command implementation and clean up imports

- Deleted the entire block command implementation from the codebase.
- Removed unused imports related to the block command in the components test file.

alvarosabu committed:
feat: add fetchComponent function and update components command to support fetching individual components

- Introduced a new `fetchComponent` function to retrieve a specific component by name.
- Updated the `componentsCommand` to allow pulling a single component by name, enhancing the command's functionality.
- Modified tests to cover the new component fetching feature, including scenarios for successful retrieval and handling of non-existent components.
- Adjusted the structure of the command's output to reflect the new functionality, ensuring clarity in success messages.

edodusi commented on src/commands/components/index.ts:
```diff
@@ -0,0 +1,94 @@
+const program = getProgram() // Get the shared singleton instance
+
+export const componentsCommand = program
+  .command('components')
```
```suggestion
  .command('components')
  .alias('comp')
```

alvarosabu committed:
Update src/commands/components/index.ts

Co-authored-by: Edoardo Dusi <edo@edoardodusi.com>

edodusi commented on src/commands/components/index.ts:
```diff
@@ -0,0 +1,94 @@
+        const fileName = suffix ? `${filename}.${suffix}.json` : `${filename}.json`
+        const filePath = path ? `${path}/${fileName}` : `.storyblok/components/${space}/${fileName}`
+
+        konsola.ok(`Components downloaded successfully in ${chalk.hex(colorPalette.PRIMARY)(filePath)}`)
```
```suggestion
        konsola.ok(`Components downloaded successfully to ${chalk.hex(colorPalette.PRIMARY)(filePath)}`)
```

alvarosabu committed:
Update src/commands/components/index.ts

Co-authored-by: Edoardo Dusi <edo@edoardodusi.com>

alvarosabu committed:
Merge branch 'next' into feature/pull-components-cmd

alvarosabu committed:
chore(tests): updated success messages in components command to use "to" instead of "in"

alvarosabu committed:
fix: update COMPONENTS color in constants and enhance components command with loading spinners

- Changed COMPONENTS color from '#FF5722' to '#a185ff' in src/constants.ts for better visual consistency.
- Added loading spinners to the components command in src/commands/components/index.ts to improve user experience during data fetching operations.
- Each data fetching step now provides visual feedback, enhancing the command's interactivity.
